### PR TITLE
Simulation store merge

### PR DESCRIPTION
**Pull request by jd-lara**
All the tests are passing locally, I opened the PR to make sure that it passes in the CI.

### CODE REVIEW

File: src/core/cache_utils.jl [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+make_cache_key(stage, type, name) = (stage = stage, type = type, name = name)
+
+# Priority for keeping data in cache to serve reads. Currently unused.
+IS.@scoped_enum CachePriority begin
+    LOW
+    MEDIUM
+    HIGH
+end
+
+# TODO: we need to make scoped_enum handle words ending in 'y'
```
Comment (kdheepak) [inline src/core/cache_utils.jl:12]: I've opened an issue to track this: https://github.com/NREL-SIIP/InfrastructureSystems.jl/issues/186

File: src/core/cache_utils.jl [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    misses::Int
+end
+
+CacheStats() = CacheStats(0, 0)
+
+function get_cache_hit_percentage(x::CacheStats)
+    total = x.hits + x.misses
+    total == 0 && return 0.0
+
+    return Float64(x.hits) / (x.hits + x.misses) * 100
```
Comment (kdheepak) [inline src/core/cache_utils.jl:54]: Can you tell me why the `Float64` is necessary here?

File: src/core/definitions.jl [hunk 3]
```diff
@@ -21,9 +21,17 @@ const JuMPParamArray = JuMP.Containers.DenseAxisArray{PJ.ParameterRef}
...
-    EMPTY = 0
+    FAILED_BUILD = 1
+    EMPTY = 2
+end
+
+@enum RUN_STATUS begin
+    READY = -1
+    SUCCESSFUL_RUN = 0
+    RUNNING = 1
+    FAILED_RUN = 2
```
Comment (kdheepak) [inline src/core/definitions.jl:34]: I have a minor preference for sticking to the same logic for BUILD_STATUS and RUN_STATUS.

```suggestion
@enum RUN_STATUS begin
    RUNNING = -1
    READY = 0
    SUCCESSFUL_RUN = 1
    FAILED_RUN = 2
```

The idea being negative numbers for current state while running (`IN_PROGRESS`/`RUNNING`), zero for the success state, and positive numbers for the state after a process is finished (`FAILED_BUILD`/`FAILED_RUN`) 

But if you have a reason for doing it the current way, I'm happy to be overruled on this.

File: src/core/definitions.jl [hunk 4]
```diff
@@ -21,9 +21,17 @@ const JuMPParamArray = JuMP.Containers.DenseAxisArray{PJ.ParameterRef}
 const DenseAxisArrayContainer = Dict{Symbol, JuMP.Containers.DenseAxisArray}
 
 @enum BUILD_STATUS begin
-    BUILT = 1
+    BUILT = 0
     IN_PROGRESS = -1
-    EMPTY = 0
+    FAILED_BUILD = 1
+    EMPTY = 2
```
Comment (kdheepak) [inline src/core/definitions.jl:27]: In increasing order:

```suggestion
@enum BUILD_STATUS begin
    IN_PROGRESS = -1
    BUILT = 0
    FAILED_BUILD = 1
    EMPTY = 2
```

File: src/core/feedforward.jl [hunk 5]
```diff
@@ -717,6 +717,7 @@ function feedforward_update!(
             IS.@record :simulation FeedForwardUpdateEvent(
                 "FeedForward",
                 current_time,
+                1,  # TODO DT
```
Comment (kdheepak) [inline src/core/feedforward.jl:720]: Can you clarify what the `TODO` is over here?

File: src/core/hdf_simulation_store.jl [hunk 6]
```diff
@@ -0,0 +1,562 @@
...
+mutable struct HdfSimulationStore <: SimulationStore
+    file::HDF5.File
+    params::SimulationStoreParams
+    # The key order is the stage execution order.
+    datasets::OrderedDict{Symbol, StageDatasets}
+    optimizer_stats_index::Int
+    cache::ResultCache
+end
+
+function HdfSimulationStore(file_path::AbstractString, mode::AbstractString)
```
Comment (kdheepak) [inline src/core/hdf_simulation_store.jl:54]: Would using `mode::Char` be more appropriate?

Comment (kdheepak) [reply src/core/hdf_simulation_store.jl:54]: Never mind, the stdlib Julia `open` uses `AbstractString`.

File: src/core/simulation.jl [hunk 7]
```diff
@@ -1,21 +1,24 @@
-
 const SIMULATION_SERIALIZATION_FILENAME = "simulation.bin"
 const SIMULATION_LOG_FILENAME = "simulation.log"
-const REQUIRED_RECORDERS = [:simulation_status]
+const REQUIRED_RECORDERS = [:simulation_status, :simulation]
```
Comment (kdheepak) [inline src/core/simulation.jl:3]: Is it possible to make this a tuple?

**Comment by daniel-thom** [comment]
> All the tests are passing locally, I opened the PR to make sure that it passes in the CI.

Not sure if Jose meant it, but I think we should consider this a draft PR. There are several TODOs that have to be addressed.

File: src/core/cache_utils.jl (hunk 2 above)
Comment (daniel-thom) [reply src/core/cache_utils.jl:54]: Doh. I don't do much math in Julia and didn't realize that `/` automatically converts to floats. Too used to Python.

File: src/core/feedforward.jl (hunk 5 above)
Comment (daniel-thom) [reply src/core/feedforward.jl:720]: This needs to be removed. I initially had an idea of tracking time step as an integer. I'll fix all instances.

File: src/core/definitions.jl (hunk 3 above)
Comment (jd-lara) [reply src/core/definitions.jl]: addressed in 22c4fcf

File: src/core/definitions.jl [hunk 8]
```diff
@@ -20,26 +20,20 @@ const JuMPVariableArray = JuMP.Containers.DenseAxisArray{JuMP.VariableRef}
 const JuMPParamArray = JuMP.Containers.DenseAxisArray{PJ.ParameterRef}
 const DenseAxisArrayContainer = Dict{Symbol, JuMP.Containers.DenseAxisArray}
 
-@enum BUILD_STATUS begin
-    BUILT = 0
+IS.@scoped_enum BUILD_STATUS begin
```
Comment (daniel-thom) [inline src/core/definitions.jl:23]: Our convention has been to make the enum type PascalCase and the enum values ALL_CAPS, like global constants.

File: src/core/definitions.jl [hunk 9]
```diff
@@ -20,26 +20,20 @@ const JuMPVariableArray = JuMP.Containers.DenseAxisArray{JuMP.VariableRef}
...
     IN_PROGRESS = -1
+    BUILT = 0
     FAILED_BUILD = 1
     EMPTY = 2
 end
 
-@enum RUN_STATUS begin
+IS.@scoped_enum RUN_STATUS begin
     READY = -1
     SUCCESSFUL_RUN = 0
```
Comment (daniel-thom) [inline src/core/definitions.jl:32]: The "RUN" is redundant. You could have READY, SUCCESSFUL, RUNNING, FAILED.

File: src/core/definitions.jl (hunk 8 above)
Comment (jd-lara) [reply src/core/definitions.jl:23]: We need to add this to the style guide then. I will make the change

File: src/core/definitions.jl (hunk 9 above)
Comment (jd-lara) [reply src/core/definitions.jl]: addressed in c6af5b0

File: src/core/definitions.jl (hunk 8 above)
Comment (jd-lara) [reply src/core/definitions.jl]: addressed in c6af5b0

**Pull request merged**